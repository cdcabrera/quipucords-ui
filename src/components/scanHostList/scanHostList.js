import React from 'react';
import PropTypes from 'prop-types';
import { EmptyState, Grid, Spinner } from 'patternfly-react';
import { connect, reduxActions, reduxSelectors } from '../../redux';
import { helpers } from '../../common/helpers';
import { apiTypes } from '../../constants/apiConstants';

class ScanHostList extends React.Component {
  state = {
    currentPage: 1,
    queryObject: {
      [apiTypes.API_QUERY_PAGE]: 1,
      [apiTypes.API_QUERY_PAGE_SIZE]: 100,
      [apiTypes.API_QUERY_ORDERING]: 'name'
    }
  };

  componentDidMount() {
    this.onRefresh();
  }

  onRefresh(updatedQueryObject = {}) {
    const { queryObject } = this.state;
    const {
      filter,
      getConnectionScanResults,
      getInspectionScanResults,
      id,
      useConnectionResults,
      useInspectionResults
    } = this.props;

    if (useConnectionResults) {
      getConnectionScanResults(id, { ...queryObject, ...filter, ...updatedQueryObject });
    }

    if (useInspectionResults) {
      getInspectionScanResults(id, { ...queryObject, ...filter, ...updatedQueryObject });
    }
  }

  onScrollList = event => {
    const { target } = event;
    const { currentPage } = this.state;
    const { isMoreResults, pending } = this.props;

    const bottom = target.scrollHeight - target.scrollTop === target.clientHeight;

    if (bottom && !pending && isMoreResults) {
      const newPage = currentPage + 1;

      const updatedQueryObject = {
        [apiTypes.API_QUERY_PAGE]: newPage
      };

      this.setState(
        {
          currentPage: newPage
        },
        () => {
          this.onRefresh(updatedQueryObject);
        }
      );
    }
  };

  render() {
    const { children, error, errorMessage, hostsList, pending } = this.props;

    if (error) {
      return (
        <EmptyState>
          <EmptyState.Icon name="error-circle-o" />
          <EmptyState.Title>Error retrieving scan results</EmptyState.Title>
          <EmptyState.Info>{errorMessage}</EmptyState.Info>
        </EmptyState>
      );
    }

    if (pending) {
      return (
        <EmptyState>
          <Spinner loading size="sm" className="blank-slate-pf-icon" />
          <EmptyState.Title>Loading...</EmptyState.Title>
        </EmptyState>
      );
    }

    return (
      <div className="quipucords-infinite-results">
        <Grid fluid onScroll={this.onScrollList} className="quipucords-infinite-list">
          {hostsList.map(host => children({ host }))}
        </Grid>
      </div>
    );
  }
}

ScanHostList.propTypes = {
  children: PropTypes.func.isRequired,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  filter: PropTypes.object,
  getConnectionScanResults: PropTypes.func,
  getInspectionScanResults: PropTypes.func,
  hostsList: PropTypes.arrayOf(
    PropTypes.shape({
      credentialName: PropTypes.string,
      jobType: PropTypes.oneOf(['connection', 'inspection']),
      name: PropTypes.string,
      sourceId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      sourceName: PropTypes.string,
      status: PropTypes.string
    })
  ),
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  isMoreResults: PropTypes.bool,
  pending: PropTypes.bool,
  useConnectionResults: PropTypes.bool,
  useInspectionResults: PropTypes.bool
};

ScanHostList.defaultProps = {
  error: false,
  errorMessage: null,
  filter: {},
  getConnectionScanResults: helpers.noop,
  getInspectionScanResults: helpers.noop,
  hostsList: [],
  isMoreResults: false,
  pending: false,
  useConnectionResults: false,
  useInspectionResults: false
};

const mapDispatchToProps = dispatch => ({
  getConnectionScanResults: (id, query) => dispatch(reduxActions.scans.getConnectionScanResults(id, query)),
  getInspectionScanResults: (id, query) => dispatch(reduxActions.scans.getInspectionScanResults(id, query))
});

const makeMapStateToProps = () => {
  const getScanHostsDetails = reduxSelectors.scans.makeScanHostsList();

  return (state, props) => ({
    ...getScanHostsDetails(state, props)
  });
};

const ConnectedScanHostList = connect(makeMapStateToProps, mapDispatchToProps)(ScanHostList);

export { ConnectedScanHostList as default, ConnectedScanHostList, ScanHostList };
