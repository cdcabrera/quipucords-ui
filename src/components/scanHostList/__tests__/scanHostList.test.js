import React from 'react';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount, shallow } from 'enzyme';
import { ConnectedScanHostList, ScanHostList } from '../scanHostList';

describe('ScanHostList Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component', () => {
    const store = generateEmptyStore({
      scans: {
        connection: {},
        inspection: {}
      }
    });

    const props = {
      id: 1,
      filter: { lorem: 'ipsum' },
      useConnectionResults: true,
      useInspectionResults: true
    };

    const component = shallow(
      <Provider store={store}>
        <ConnectedScanHostList {...props}>{() => 'lorem ipsum'}</ConnectedScanHostList>
      </Provider>
    );

    expect(component.find(ConnectedScanHostList)).toMatchSnapshot('connected');
  });

  it('should render a non-connected component', () => {
    const props = {
      id: 1,
      useConnectionResults: true,
      useInspectionResults: true,
      hostsList: [
        {
          credentialName: 'dolor',
          jobType: 'connection',
          name: 'lorem',
          sourceId: 15,
          sourceName: 'lorem source'
        },
        {
          credentialName: 'set',
          jobType: 'inspection',
          name: 'ipsum',
          sourceId: 16,
          sourceName: 'ipsum source'
        }
      ]
    };

    const component = mount(<ScanHostList {...props}>{({ host }) => JSON.stringify(host)}</ScanHostList>);

    expect(component.render()).toMatchSnapshot('non-connected');
  });

  it('should render a non-connected component error', () => {
    const props = {
      id: 1,
      error: true,
      errorMessage: 'Lorem Ipsum.'
    };

    const component = shallow(<ScanHostList {...props}>{({ host }) => JSON.stringify(host)}</ScanHostList>);
    expect(component).toMatchSnapshot('error');
  });

  it('should render a non-connected component pending', () => {
    const props = {
      id: 1,
      pending: true
    };

    const component = shallow(<ScanHostList {...props}>{({ host }) => JSON.stringify(host)}</ScanHostList>);
    expect(component).toMatchSnapshot('pending');
  });
});
