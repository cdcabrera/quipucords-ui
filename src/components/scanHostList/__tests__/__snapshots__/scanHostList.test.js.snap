// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanHostList Component should render a connected component: connected 1`] = `
<ScanHostList
  error={false}
  errorMessage={null}
  filter={
    Object {
      "lorem": "ipsum",
    }
  }
  getConnectionScanResults={[Function]}
  getInspectionScanResults={[Function]}
  hostsList={Array []}
  id={1}
  isMoreResults={false}
  pending={false}
  useConnectionResults={true}
  useInspectionResults={true}
>
  <Component />
</ScanHostList>
`;

exports[`ScanHostList Component should render a non-connected component error: error 1`] = `
<EmptyState
  className=""
>
  <EmptyStateIcon
    className=""
    name="error-circle-o"
    type="pf"
  />
  <EmptyStateTitle
    className=""
  >
    Error retrieving scan results
  </EmptyStateTitle>
  <EmptyStateInfo
    className=""
  >
    Lorem Ipsum.
  </EmptyStateInfo>
</EmptyState>
`;

exports[`ScanHostList Component should render a non-connected component pending: pending 1`] = `
<EmptyState
  className=""
>
  <Spinner
    className="blank-slate-pf-icon"
    inline={false}
    inverse={false}
    loading={true}
    size="sm"
  />
  <EmptyStateTitle
    className=""
  >
    Loading...
  </EmptyStateTitle>
</EmptyState>
`;

exports[`ScanHostList Component should render a non-connected component: non-connected 1`] = `
<div
  class="quipucords-infinite-results"
>
  <div
    class="quipucords-infinite-list container-fluid"
  >
    {"credentialName":"dolor","jobType":"connection","name":"lorem","sourceId":15,"sourceName":"lorem source"}{"credentialName":"set","jobType":"inspection","name":"ipsum","sourceId":16,"sourceName":"ipsum source"}
  </div>
</div>
`;
